File: Samples/Geolocation/cs/BackgroundTask/VisitBackgroundTask.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using Windows.ApplicationModel.Background;
using Windows.Data.Json;
using Windows.Data.Xml.Dom;
using Windows.Devices.Geolocation;
using Windows.Globalization;
using Windows.Globalization.DateTimeFormatting;
using Windows.Storage;
using Windows.UI.Notifications;

namespace BackgroundTask
{
    public sealed class VisitBackgroundTask : IBackgroundTask
    {
        private ObservableCollection<string> _visitBackgroundEvents = new ObservableCollection<string>();
        private const int maxEventDescriptors = 42;                     // Value determined by how many max length event descriptors (91 chars) 
                                                                        // stored as a JSON string can fit in 8K (max allowed for local settings)

        void IBackgroundTask.Run(IBackgroundTaskInstance taskInstance)
        {
            BackgroundTaskDeferral deferral = taskInstance.GetDeferral();
            GeovisitTriggerDetails triggerDetails = (GeovisitTriggerDetails)taskInstance.TriggerDetails;

            // Handle Visit reports
            GetVisitReports(triggerDetails);
            deferral.Complete();
        }

        private void GetVisitReports(GeovisitTriggerDetails triggerDetails)
        {
            // First fill the _visitBackgroundEvents with previous existing reports saved in local settings.
            _visitBackgroundEvents.Clear();
            FillReportCollectionWithExistingReports();

            // Read reports from the triggerDetails.
            IReadOnlyList<Geovisit> reports = triggerDetails.ReadReports();

            string visitItemEvent = null;
            int numReports = reports.Count;

            DateTimeFormatter formatterLongTime = new DateTimeFormatter("{hour.integer}:{minute.integer(2)}:{second.integer(2)}", new[] { "en-US" }, "US", CalendarIdentifiers.Gregorian, ClockIdentifiers.TwentyFourHour);

            foreach (Geovisit report in reports)
            {
                visitItemEvent = formatterLongTime.Format(report.Timestamp);

                visitItemEvent += " " + report.StateChange.ToString();

                // Check if the report has a valid position.
                if (report.Position != null)
                {
                    visitItemEvent += " (" +
                    report.Position.Coordinate.Point.Position.Latitude.ToString() + "," +
                    report.Position.Coordinate.Point.Position.Longitude.ToString() +
                    ")";
                }

                // Now add each new visit report to _VisitBackgroundEvents.
                AddVisitReport(visitItemEvent);
            }

            if (numReports != 0)
            {
                // Save all the reports back to the local settings.
                SaveExistingReports();

                // NOTE: Other notification mechanisms can be used here, such as Badge and/or Tile updates.
                DoToast(numReports, visitItemEvent);
            }

        }

        /// <summary>
        /// Helper method to pop a toast
        /// </summary>
        private void DoToast(int numEventsOfInterest, string eventName)
        {
            // pop a toast for each visit event
            ToastNotifier ToastNotifier = ToastNotificationManager.CreateToastNotifier();

            // Create a two line toast and add audio reminder

            // Here the xml that will be passed to the 
            // ToastNotification for the toast is retrieved
            XmlDocument toastXml = ToastNotificationManager.GetTemplateContent(ToastTemplateType.ToastText02);

            // Set both lines of text
            XmlNodeList toastNodeList = toastXml.GetElementsByTagName("text");
            toastNodeList.Item(0).AppendChild(toastXml.CreateTextNode("Geolocation Sample"));

            if (1 == numEventsOfInterest)
            {
                toastNodeList.Item(1).AppendChild(toastXml.CreateTextNode(eventName));
            }
            else
            {
                string secondLine = "There are " + numEventsOfInterest + " new visit events";
                toastNodeList.Item(1).AppendChild(toastXml.CreateTextNode(secondLine));
            }

            // now create a xml node for the audio source
            IXmlNode toastNode = toastXml.SelectSingleNode("/toast");
            XmlElement audio = toastXml.CreateElement("audio");
            audio.SetAttribute("src", "ms-winsoundevent:Notification.SMS");

            ToastNotification toast = new ToastNotification(toastXml);
            ToastNotifier.Show(toast);
        }

        private void FillReportCollectionWithExistingReports()
        {
            var settings = ApplicationData.Current.LocalSettings;
            if (settings.Values.ContainsKey("BackgroundVisitEventCollection"))
            {
                string visitEvent = settings.Values["BackgroundVisitEventCollection"].ToString();

                if (0 != visitEvent.Length)
                {
                    var events = JsonValue.Parse(visitEvent).GetArray();

                    // NOTE: the events are accessed in reverse order
                    // because the events were added to JSON from
                    // newer to older.  AddVisitReport() adds
                    // each new entry to the beginning of the collection
                    // and throws away the last entry if the collection gets too large.
                    for (int pos = events.Count - 1; pos >= 0; pos--)
                    {
                        var element = events.GetStringAt((uint)pos);
                        AddVisitReport(element);
                    }
                }
            }
        }

        private void SaveExistingReports()
        {
            var jsonArray = new JsonArray();

            foreach (var eventDescriptor in _visitBackgroundEvents)
            {
                jsonArray.Add(JsonValue.CreateStringValue(eventDescriptor.ToString()));
            }

            string jsonString = jsonArray.Stringify();

            var settings = ApplicationData.Current.LocalSettings;
            settings.Values["BackgroundVisitEventCollection"] = jsonString;
        }

        private void AddVisitReport(string eventDescription)
        {
            if (_visitBackgroundEvents.Count == maxEventDescriptors)
            {
                _visitBackgroundEvents.RemoveAt(maxEventDescriptors - 1);
            }

            _visitBackgroundEvents.Insert(0, eventDescription);
        }

    }
}

File: Samples/Geolocation/cpp/BackgroundTask/VisitBackgroundTask.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

// VisitBackgroundTask.cpp
#include "pch.h"
#include "VisitBackgroundTask.h"

using namespace BackgroundTask;

using namespace Concurrency;
using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::ApplicationModel::Background;
using namespace Windows::Data::Json;
using namespace Windows::Data::Xml::Dom;
using namespace Windows::Devices::Geolocation;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::Globalization;
using namespace Windows::Globalization::DateTimeFormatting;
using namespace Windows::Storage;
using namespace Windows::UI::Notifications;

const unsigned int maxEventDescriptors = 42;    // Value determined by how many max length event descriptors (91 chars) 
                                                // stored as a JSON string can fit in 8K (max allowed for local settings)

VisitBackgroundTask::VisitBackgroundTask()
    : _visitBackgroundEvents(ref new Vector<String^>())
{
}

VisitBackgroundTask::~VisitBackgroundTask()
{
}

void VisitBackgroundTask::Run(IBackgroundTaskInstance^ taskInstance)
{
    // Get the deferral object from the task instance
    BackgroundTaskDeferral^ deferral = taskInstance->GetDeferral();

    GeovisitTriggerDetails^ triggerDetails = safe_cast<GeovisitTriggerDetails^>(taskInstance->TriggerDetails);

    GetVisitReports(triggerDetails);

    // Indicate that the background task has completed
    deferral->Complete();
}


// Gets the new visits reports and saves them into local settings
void VisitBackgroundTask::GetVisitReports(GeovisitTriggerDetails^ triggerDetails)
{
     // First fill the _visitBackgroundEvents with previous existing reports saved in local settings.
    _visitBackgroundEvents->Clear();
    FillReportCollectionWithExistingReports();

    // Read reports from the triggerDetails.
    // If the user revoked location access, ReadReports returns 0 reports.
    IVectorView<Geovisit^>^ reports = triggerDetails->ReadReports();

    auto us = ref new Vector<String^>({ "en-US" });
    auto formatterLongTime = ref new DateTimeFormatter("{hour.integer}:{minute.integer(2)}:{second.integer(2)}", us, "US", CalendarIdentifiers::Gregorian, ClockIdentifiers::TwentyFourHour);

    unsigned int count = reports->Size;

    String^ visitItemEvent = nullptr;
  
    for (unsigned int index = 0; index < count; index++)
    {
        Geovisit^ report = reports->GetAt(index);

        visitItemEvent = formatterLongTime->Format(report->Timestamp);

        visitItemEvent += " " + report->StateChange.ToString();

        // Check if the report has a valid position.
        if (report->Position != nullptr)
        {
            Geoposition^ pos = report->Position;

            visitItemEvent += " (" +
              pos->Coordinate->Point->Position.Latitude.ToString() + "," +
              pos->Coordinate->Point->Position.Longitude.ToString() +
             ")";
        }

        // Now add each new visit report to _VisitBackgroundEvents.
        AddVisitReport(visitItemEvent);
    }

    if (0 != count)
    {
        // Save all the reports back to the local settings.
        SaveExistingEvents();

        // NOTE: Other notification mechanisms can be used here, such as Badge and/or Tile updates.
        DoToast(count, visitItemEvent);
    }
}

// Helper method to pop a toast
void VisitBackgroundTask::DoToast(int numEventsOfInterest, String^ eventName)
{
    // pop a toast for each geofence event
    ToastNotifier^ ToastNotifier = ToastNotificationManager::CreateToastNotifier();

    // Create a two line toast and add audio reminder

    // Here the xml that will be passed to the 
    // ToastNotification for the toast is retrieved
    XmlDocument^ toastXml = ToastNotificationManager::GetTemplateContent(ToastTemplateType::ToastText02);

    // Set both lines of text
    XmlNodeList^ nodeList = toastXml->GetElementsByTagName("text");
    nodeList->Item(0)->AppendChild(toastXml->CreateTextNode("Geolocation Sample"));

    if (1 == numEventsOfInterest)
    {
        nodeList->Item(1)->AppendChild(toastXml->CreateTextNode(eventName));
    }
    else
    {
        auto secondLine = "There are " + numEventsOfInterest + " new visit events";
        nodeList->Item(1)->AppendChild(toastXml->CreateTextNode(secondLine));
    }

    // now create a xml node for the audio source
    IXmlNode^ toastNode = toastXml->SelectSingleNode("/toast");
    XmlElement^ audio = toastXml->CreateElement("audio");
    audio->SetAttribute("src", "ms-winsoundevent:Notification.SMS");

    ToastNotification^ toast = ref new ToastNotification(toastXml);
    ToastNotifier->Show(toast);
}

void VisitBackgroundTask::FillReportCollectionWithExistingReports()
{
    auto settings = ApplicationData::Current->LocalSettings->Values;
    if (settings->HasKey("BackgroundVisitEventCollection"))
    {
        String^ visitEvent = safe_cast<String^>(settings->Lookup("BackgroundVisitEventCollection"));

        if (0 != visitEvent->Length())
        {
            auto events = JsonValue::Parse(visitEvent)->GetArray();

            // NOTE: the events are accessed in reverse order
            // because the events were added to JSON from
            // newer to older.  AddVisitReport() adds
            // each new entry to the beginning of the collection
            // and throws away the last entry if the collection gets too large.
            for (int pos = events->Size - 1; pos >= 0; pos--)
            {
                auto element = events->GetAt(pos);
                AddVisitReport(element->GetString());
            }
        }
    }
}

void VisitBackgroundTask::SaveExistingEvents()
{
    auto jsonArray = ref new JsonArray();

    for (auto eventDescriptor : _visitBackgroundEvents)
    {
        jsonArray->Append(JsonValue::CreateStringValue(eventDescriptor->ToString()));
    }

    String^ jsonString = jsonArray->Stringify();

    auto settings = ApplicationData::Current->LocalSettings->Values;
    settings->Insert("BackgroundVisitEventCollection", jsonString);
}

void VisitBackgroundTask::AddVisitReport(String^ eventDescription)
{
    if (_visitBackgroundEvents->Size == maxEventDescriptors)
    {
        _visitBackgroundEvents->RemoveAtEnd();
    }

    _visitBackgroundEvents->InsertAt(0, eventDescription);
}
